### PR TITLE
Letli/pks 3374/coredns pods resillient

### PR DESCRIPTION
**Pull request by 996refuse**
**What this PR does / why we need it**: https://jira.eng.vmware.com/browse/PKS-3374

**How can this PR be verified?**

**Is there any change in kubo-deployment?**

**Is there any change in kubo-ci?**

**Does this affect upgrade, or is there any migration required?**

**Which issue(s) this PR fixes:**

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by ahrtr** [comment]
@996refuse A couple of comments:
1. Please double confirm that each worker node has the label "topology.kubernetes.io/zone", and its value should be identical to "failure-domain.beta.kubernetes.io/zone".
2. Please add at least one test case to verify the 3 CoreDNS POD spread across the zones & node evenly;
3. Since you changed the pks-kubernetes-release repo, so please you need to run a cfcr pipeline firstly, please refer to https://pks.ci.cf-app.com/teams/dev/pipelines/cfcr-dependencies-1.10.x. @YanzhaoLi and @qing-chi-2013 have more knowledge on how to run cfcr pipeline. 
4. Please paste the pipeline link into this PR once it passes;
5. Can you please clearly state the manually tests & test results you have done so far in this PR? 

@wjun Based on your comment in https://jira.eng.vmware.com/browse/PKS-3374, you only suggested to add label "topology.kubernetes.io/zone" for worker node on vSphere IaaS. Should we support the public cloud environments either, such as AWS, Azure? I see that the existing label "failure-domain.beta.kubernetes.io/zone" was only applied to vSphere, I don't get why it wasn't applied to public cloud?

File: jobs/kubelet/templates/bin/kubelet_ctl.erb [hunk 1]
```diff
@@ -41,6 +41,7 @@ kubectl="/var/vcap/packages/kubernetes/bin/kubectl --kubeconfig=/var/vcap/jobs/k
 
   if iaas=="vsphere"
     labels << "failure-domain.beta.kubernetes.io/zone=#{spec.az}"
+    labels << "topology.kubernetes.io/zone=#{spec.az}"
```
Comment (ahrtr) [inline jobs/kubelet/templates/bin/kubelet_ctl.erb:44]: Should we apply the labels to public cloud, such as AWS, Azure?

File: jobs/apply-specs/templates/specs/coredns.yml.erb [hunk 2]
```diff
@@ -100,6 +100,19 @@ spec:
...
+          matchLabels:
+            k8s-app: kube-dns
+        maxSkew: 1
+        topologyKey: topology.kubernetes.io/zone
+        whenUnsatisfiable: ScheduleAnyway
+      - labelSelector:
+          matchLabels:
+            k8s-app: kube-dns
+        maxSkew: 1
+        topologyKey: kubernetes.io/hostname
```
Comment (DanielXiao) [inline jobs/apply-specs/templates/specs/coredns.yml.erb:114]: Use unified label domain name `topology.kubernetes.io/hostname`. You need to label each node with this key.

File: jobs/kubelet/templates/bin/kubelet_ctl.erb (hunk 1 above)
Comment (DanielXiao) [reply jobs/kubelet/templates/bin/kubelet_ctl.erb:44]: Yes, we should.

File: jobs/apply-specs/templates/specs/coredns.yml.erb (hunk 2 above)
Comment (996refuse) [reply jobs/apply-specs/templates/specs/coredns.yml.erb:114]: @DanielXiao you mean `topology.kubernetes.io/zone` and `topology.kubernetes.io/hostname`are unified right? `kubernetes.io/hostname` is from doc https://kubernetes.io/zh/docs/concepts/workloads/pods/pod-topology-spread-constraints/#internal-default-constraints i think follow the official doc is better..

Comment (DanielXiao) [reply jobs/apply-specs/templates/specs/coredns.yml.erb:114]: Stick to kubernetes.io/hostname ,  I was not aware that it is the builtin topologykey.

**Comment by 996refuse** [comment]
cfcr pipeline
https://pks.ci.cf-app.com/teams/bosh-lifecycle-dev/pipelines/cfcr-dependencies-1.10.x.letli.coredns-coredns-1.10.x

pks pipeline
https://pks.ci.cf-app.com/teams/bosh-lifecycle-dev/pipelines/pks-api-1.10.x.letli.coredns-coredns-1.10.x

**Comment by 996refuse** [comment]
hi @yuyangbj @ahrtr 
i have installed the latest pks(1.10.0-build.22) on aws and azure, it shows that `topology.kubernetes.io/zone` is automatically added by the platforms, so only need to add the label for vsphere. please help double confirm.

aws
```
ubuntu@ip-10-0-0-240:~$ kubectl get nodes --show-labels                                                     
NAME                                      STATUS   ROLES    AGE   VERSION            LABELS
ip-10-0-10-4.us-east-2.compute.internal   Ready    <none>   52m   v1.19.6+vmware.1   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m5.large,beta.kubernetes.io/os=linux,bosh.id=b316b447-5620-4974-95d5-b20e12064b2b,bosh.zone=us-east-2c,failure-domain.beta.kubernetes.io/region=us-east-2,failure-domain.beta.kubernetes.io/zone=us-east-2c,kubernetes.io/arch=amd64,kubernetes.io/hostname=10.0.10.4,kubernetes.io/os=linux,node.kubernetes.io/instance-type=m5.large,pks-system/cluster.name=foo,pks-system/cluster.uuid=service-instance_9b37dfe0-297c-43d1-8c6b-132fad0223ff,spec.ip=10.0.10.4,topology.kubernetes.io/region=us-east-2,topology.kubernetes.io/zone=us-east-2c
ip-10-0-8-5.us-east-2.compute.internal    Ready    <none>   56m   v1.19.6+vmware.1   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m5.large,beta.kubernetes.io/os=linux,bosh.id=e4d8085b-8d10-47cf-be18-f68bd8d4f40b,bosh.zone=us-east-2a,failure-domain.beta.kubernetes.io/region=us-east-2,failure-domain.beta.kubernetes.io/zone=us-east-2a,kubernetes.io/arch=amd64,kubernetes.io/hostname=10.0.8.5,kubernetes.io/os=linux,node.kubernetes.io/instance-type=m5.large,pks-system/cluster.name=foo,pks-system/cluster.uuid=service-instance_9b37dfe0-297c-43d1-8c6b-132fad0223ff,spec.ip=10.0.8.5,topology.kubernetes.io/region=us-east-2,topology.kubernetes.io/zone=us-east-2a
ip-10-0-9-4.us-east-2.compute.internal    Ready    <none>   54m   v1.19.6+vmware.1   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m5.large,beta.kubernetes.io/os=linux,bosh.id=9b42d408-7441-4388-99cf-a41f859e7298,bosh.zone=us-east-2b,failure-domain.beta.kubernetes.io/region=us-east-2,failure-domain.beta.kubernetes.io/zone=us-east-2b,kubernetes.io/arch=amd64,kubernetes.io/hostname=10.0.9.4,kubernetes.io/os=linux,node.kubernetes.io/instance-type=m5.large,pks-system/cluster.name=foo,pks-system/cluster.uuid=service-instance_9b37dfe0-297c-43d1-8c6b-132fad0223ff,spec.ip=10.0.9.4,topology.kubernetes.io/region=us-east-2,topology.kubernetes.io/zone=us-east-2b
```

azure
```
ubuntu@cosmic-piper-ops-manager:~$ kubectl get nodes --show-labels
NAME                                   STATUS   ROLES    AGE   VERSION            LABELS
2469ceec-8c58-474e-af0c-44d8496076ec   Ready    <none>   28m   v1.19.6+vmware.1   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=Standard_F2s,beta.kubernetes.io/os=linux,bosh.id=c638aee0-67e4-447a-8b43-0189d0250b16,bosh.zone=zone-3,failure-domain.beta.kubernetes.io/region=westus2,failure-domain.beta.kubernetes.io/zone=westus2-3,kubernetes.io/arch=amd64,kubernetes.io/hostname=2469ceec-8c58-474e-af0c-44d8496076ec,kubernetes.io/os=linux,node.kubernetes.io/instance-type=Standard_F2s,pks-system/cluster.name=foo,pks-system/cluster.uuid=service-instance_31612a98-df1e-4613-ad73-bdf402da2071,spec.ip=10.0.16.7,topology.kubernetes.io/region=westus2,topology.kubernetes.io/zone=westus2-3
61b156f1-ba38-4e98-b058-a0573c220df7   Ready    <none>   32m   v1.19.6+vmware.1   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=Standard_F2s,beta.kubernetes.io/os=linux,bosh.id=751e3616-fb4f-4129-bc60-00a0f7b8547d,bosh.zone=zone-2,failure-domain.beta.kubernetes.io/region=westus2,failure-domain.beta.kubernetes.io/zone=westus2-2,kubernetes.io/arch=amd64,kubernetes.io/hostname=61b156f1-ba38-4e98-b058-a0573c220df7,kubernetes.io/os=linux,node.kubernetes.io/instance-type=Standard_F2s,pks-system/cluster.name=foo,pks-system/cluster.uuid=service-instance_31612a98-df1e-4613-ad73-bdf402da2071,spec.ip=10.0.16.6,topology.kubernetes.io/region=westus2,topology.kubernetes.io/zone=westus2-2
8a4a376d-067a-4404-9ad7-93d915465e80   Ready    <none>   35m   v1.19.6+vmware.1   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=Standard_F2s,beta.kubernetes.io/os=linux,bosh.id=e10584a0-32e9-4873-ada3-d6decb512b53,bosh.zone=zone-1,failure-domain.beta.kubernetes.io/region=westus2,failure-domain.beta.kubernetes.io/zone=westus2-1,kubernetes.io/arch=amd64,kubernetes.io/hostname=8a4a376d-067a-4404-9ad7-93d915465e80,kubernetes.io/os=linux,node.kubernetes.io/instance-type=Standard_F2s,pks-system/cluster.name=foo,pks-system/cluster.uuid=service-instance_31612a98-df1e-4613-ad73-bdf402da2071,spec.ip=10.0.16.5,topology.kubernetes.io/region=westus2,topology.kubernetes.io/zone=westus2-1
```

**Review by ahrtr — APPROVED** [review]

**Comment by ahrtr** [comment]
@996refuse you need to submit a PR to integrate your change into p-pks-integration as well. Please sync with @qing-chi-2013  & @YanzhaoLi .

**Pull request merged**